src/Ztm.WebApi/TransactionConfirmationWatcher.cs

```diff
@@ -0,0 +1,404 @@
+        {
+            await this.ruleRepository.UpdateStatusAsync(watch.Context.Id, TransactionConfirmationWatchingRuleStatus.Success, cancellationToken);
+            await ExecuteCallbackAsync(watch.Context.Callback, watch.Context.Success, CancellationToken.None);
+        }
+
+        async Task ExecuteCallbackAsync(Callback callback, TransactionConfirmationCallbackResult payload, CancellationToken cancellationToken)
+        {
+            await this.callbackRepository.AddHistoryAsync(callback.Id, payload, cancellationToken);
+
+            try
+            {
+                await this.callbackExecuter.Execute(callback.Id, callback.Url, payload);
+            }
+            catch (Exception ex) // lgtm [cs/catch-of-all-exceptions]
+            {
+                this.logger.LogError($"Callback execution is fail.", ex);
+                return;
+            }
+
+            await this.callbackRepository.SetCompletedAsyc(callback.Id, CancellationToken.None);
+        }
+
+        Task IBlockListener.BlockAddedAsync(Block block, int height, CancellationToken cancellationToken)
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Added, cancellationToken);
+        }
+
+        Task IBlockListener.BlockRemovingAsync(Block block, int height, CancellationToken cancellationToken)
+        {
+            return this.watcher.ExecuteAsync(block, height, BlockEventType.Removing, cancellationToken);
+        }
+
+        async Task<IEnumerable<Rule>> ITransactionConfirmationWatcherHandler<Rule>.CreateContextsAsync(Transaction tx, CancellationToken cancellationToken)
+        {
+            this.timerLock.EnterReadLock();
+
+            try
+            {
+                if (this.timers.TryGetValue(tx.GetHash(), out var txTimers) && txTimers.Count > 0)
+                {
+                    var contexts = new Collection<Rule>();
+
+                    foreach (var timer in txTimers)
+                    {
+                        var rule = await this.ruleRepository.GetAsync(timer.Key, cancellationToken);
+                        contexts.Add(rule);
+                    }
+
+                    return contexts;
+                }
+            }
+            finally
+            {
+                this.timerLock.ExitReadLock();
+            }
+
+            return Enumerable.Empty<Rule>();
+        }
+
+        async Task<bool> IConfirmationWatcherHandler<TransactionWatch<Rule>, Rule>.ConfirmationUpdateAsync(TransactionWatch<Rule> watch, int confirmation, ConfirmationType type, CancellationToken cancellationToken)
+        {
+            switch (type)
+            {
+                case ConfirmationType.Confirmed:
+                    if (confirmation >= watch.Context.Confirmation)
+                    {
+                        await ConfirmAsync(watch, cancellationToken);
+                        return true;
+                    }
+                    break;
+                case ConfirmationType.Unconfirming:
```

Comment:
We should remove current watch from rule here when `confirmation` is `1`.

Reply:
Added logic to remove watch to the tail of remove watch method.